### PR TITLE
mpid/common/shm: Fix length of serialized handle copy

### PR DESCRIPTION
**Pull request by raffenet**
Avoid reading beyond the actual length of the serialized handle.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by raffenet** [comment]
Found this bug while testing out address sanitizer on MPICH:
```
Unexpected output in attric: =================================================================
Unexpected output in attric: ==372003==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60600003971f at pc 0x7fb3ebd64366 bp 0x7fffd53089b0 sp 0x7fffd53089a8
Unexpected output in attric: READ of size 256 at 0x60600003971f thread T0
Unexpected output in attric:     #0 0x7fa3ad70c365 in MPIDU_Init_shm_put /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/common/shm/mpidu_init_shm.c:277
Unexpected output in attric:     #1 0x7fa3ad70af2c in MPIDU_shm_seg_alloc /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/common/shm/mpidu_shm_alloc.c:110
Unexpected output in attric:     #0 0x7fb3ebd64365 in MPIDU_Init_shm_put /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/common/shm/mpidu_init_shm.c:277
Unexpected output in attric:     #1 0x7fb3ebd62f2c in MPIDU_shm_seg_alloc /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/common/shm/mpidu_shm_alloc.c:110
Unexpected output in attric:     #2 0x7fa3ad6dd866 in MPID_nem_init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c:252
Unexpected output in attric:     #3 0x7fa3ad6b5872 in MPIDI_CH3_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/channels/nemesis/src/ch3_init.c:95
Unexpected output in attric:     #4 0x7fa3ad693b64 in MPID_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/src/mpid_init.c:181
Unexpected output in attric:     #5 0x7fa3ad40d42c in MPIR_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpi/init/initthread.c:125
Unexpected output in attric:     #6 0x7fa3ad40d02a in MPI_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpi/init/initthread.c:240
Unexpected output in attric:     #2 0x7fb3ebd35866 in MPID_nem_init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/channels/nemesis/src/mpid_nem_init.c:252
Unexpected output in attric:     #3 0x7fb3ebd0d872 in MPIDI_CH3_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/channels/nemesis/src/ch3_init.c:95
Unexpected output in attric:     #4 0x7fb3ebcebb64 in MPID_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/ch3/src/mpid_init.c:181
Unexpected output in attric:     #5 0x7fb3eba6542c in MPIR_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpi/init/initthread.c:125
Unexpected output in attric:     #6 0x7fb3eba6502a in MPI_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpi/init/initthread.c:240
Unexpected output in attric:     #7 0x47ef07 in MTest_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/util/mtest.c:161
Unexpected output in attric:     #8 0x480020 in MTest_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/util/mtest.c:244
Unexpected output in attric:     #9 0x47cc9b in main /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/attr/attric.c:32
Unexpected output in attric:     #7 0x47ef07 in MTest_Init_thread /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/util/mtest.c:161
Unexpected output in attric:     #8 0x480020 in MTest_Init /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/util/mtest.c:244
Unexpected output in attric:     #9 0x47cc9b in main /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/attr/attric.c:32
Unexpected output in attric:     #10 0x7fa3ac297494 in __libc_start_main (/lib64/libc.so.6+0x22494)
Unexpected output in attric:     #11 0x47c97c in _start (/var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/attr/attric+0x47c97c)
Unexpected output in attric: 
Unexpected output in attric: 0x60600003971f is located 5 bytes to the right of 58-byte region [0x6060000396e0,0x60600003971a)
Unexpected output in attric: allocated by thread T0 here:
Unexpected output in attric:     #10 0x7fb3ea8ef494 in __libc_start_main (/lib64/libc.so.6+0x22494)
Unexpected output in attric:     #11 0x47c97c in _start (/var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/attr/attric+0x47c97c)
Unexpected output in attric: SUMMARY: AddressSanitizer: heap-buffer-overflow /var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/src/mpid/common/shm/mpidu_init_shm.c:277 MPIDU_Init_shm_put
Unexpected output in attric: Shadow bytes around the buggy address:
Unexpected output in attric:   0x0c0c7ffff290: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff2a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff2b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff2c0: fa fa fa fa 00 00 00 00 00 00 02 fa fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff2d0: 00 00 00 00 00 00 02 fa fa fa fa fa 00 00 00 00
Unexpected output in attric: =>0x0c0c7ffff2e0: 00 00 00[02]fa fa fa fa fd fd fd fd fd fd fd fa
Unexpected output in attric:   0x0c0c7ffff2f0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff300: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
Unexpected output in attric:   0x0c0c7ffff310: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
Unexpected output in attric:   0x0c0c7ffff320: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
Unexpected output in attric:   0x0c0c7ffff330: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
Unexpected output in attric: Shadow byte legend (one shadow byte represents 8 application bytes):
Unexpected output in attric:   Addressable:           00
Unexpected output in attric:   Partially addressable: 01 02 03 04 05 06 07 
Unexpected output in attric:   Heap left redzone:     fa
Unexpected output in attric:   Heap right redzone:    fb
Unexpected output in attric:   Freed heap region:     fd
Unexpected output in attric:   Stack left redzone:    f1
Unexpected output in attric:   Stack mid redzone:     f2
Unexpected output in attric:   Stack right redzone:   f3
Unexpected output in attric:   Stack partial redzone: f4
Unexpected output in attric:   Stack after return:    f5
Unexpected output in attric:   Stack use after scope: f8
Unexpected output in attric:   Global redzone:        f9
Unexpected output in attric:   Global init order:     f6
Unexpected output in attric:   Poisoned by user:      f7
Unexpected output in attric:   ASan internal:         fe
Unexpected output in attric: ==372002==ABORTING
Unexpected output in attric:     #0 0x466899 in malloc (/var/lib/jenkins-slave/workspace/mpich-master-ch3-asan/compiler/clang/jenkins_configure/asan/label/centos64/mpich-master/test/mpi/attr/attric+0x466899)
Unexpected output in attric:     #1 0x7fb3ea959249 in __GI___strdup (/lib64/libc.so.6+0x8c249)
```

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
The max shm filename macros do not match, leading to this problem:
```
src/mpid/common/shm/mpidu_shm.h:#define MPIDU_SHM_MAX_FNAME_LEN 256
src/mpl/include/mpl_shm_mmap.h:#define MPL_SHM_FNAME_LEN      50
```

**Comment by raffenet** [comment]
@hzhou

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 1]
```diff
@@ -107,7 +107,7 @@ int MPIDU_shm_seg_alloc(size_t len, void **ptr)
             serialized_hnd_size = strlen(serialized_hnd);
             MPIR_Assert(serialized_hnd_size < MPIDU_SHM_MAX_FNAME_LEN);
 
-            MPIDU_Init_shm_put(serialized_hnd, MPIDU_SHM_MAX_FNAME_LEN);
+            MPIDU_Init_shm_put(serialized_hnd, serialized_hnd_size);
```
Comment (hzhou) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Yes, we need +1 to account the 0-terminator. We also need remove the `MPIR_Assert`. There is an assert inside `MPIDU_Init_shm_put` anyway.

Comment (raffenet) [reply src/mpid/common/shm/mpidu_shm_alloc.c]: OK, will fix.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Pull request merged**